sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java

```diff
@@ -327,8 +372,14 @@ public boolean advance() throws IOException {
       if (Long.MAX_VALUE - source.stride < current) {
         return false;
       }
+      Instant nextTimeToProduce = lastTimeProduced.plus(source.period);
+      if (Instant.now().isBefore(nextTimeToProduce)) {
+        return false;
+      }
       current += source.stride;
       currentTimestamp = source.timestampFn.apply(current);
+      // Advance the time by at most the period, so if we're behind we can catch up.
+      lastTimeProduced = nextTimeToProduce;
```

Comment:
That's fine; this is explicitly an overall rate and not a "produce inputs with this much time between each element"